engine/access/state_stream/backend/backend_executiondata.go

```diff
@@ -32,26 +34,59 @@ type ExecutionDataBackend struct {
 
 	subscriptionHandler  *subscription.SubscriptionHandler
 	executionDataTracker tracker.ExecutionDataTracker
+
+	executionResultProvider optimistic_sync.ExecutionResultInfoProvider
+	executionStateCache     optimistic_sync.ExecutionStateCache
 }
 
-func (b *ExecutionDataBackend) GetExecutionDataByBlockID(ctx context.Context, blockID flow.Identifier) (*execution_data.BlockExecutionData, error) {
-	header, err := b.headers.ByBlockID(blockID)
+// GetExecutionDataByBlockID retrieves execution data for a specific block by its block ID.
+//
+// CAUTION: this layer SIMPLIFIES the ERROR HANDLING convention
+//   - All errors returned are guaranteed to be benign. The node can continue normal operations after such errors.
+//   - To prevent delivering incorrect results to clients in case of an error, all other return values should be discarded.
+//
+// Expected errors:
+// - [access.DataNotFoundError]: when data required to process the request is not available.
+func (b *ExecutionDataBackend) GetExecutionDataByBlockID(
+	ctx context.Context,
+	blockID flow.Identifier,
+	criteria optimistic_sync.Criteria,
+) (*execution_data.BlockExecutionData, *accessmodel.ExecutorMetadata, error) {
+	execResultInfo, err := b.executionResultProvider.ExecutionResultInfo(blockID, criteria)
 	if err != nil {
-		return nil, fmt.Errorf("could not get block header for %s: %w", blockID, err)
+		err = access.RequireErrorIs(ctx, err, storage.ErrNotFound)
```

Comment:
this may also return another sentinel
https://github.com/onflow/flow-go/pull/7952/files#r2437800241